cmd/controller/app/options/options.go

```diff
@@ -241,14 +248,43 @@ func (s *ControllerOptions) AddFlags(fs *pflag.FlagSet) {
 			"DNS01 check requests. This should be a list containing IP address and "+
 			"port, for example 8.8.8.8:53,8.8.4.4:53")
 	fs.MarkDeprecated("dns01-self-check-nameservers", "Deprecated in favour of dns01-recursive-nameservers")
+
 	fs.BoolVar(&s.EnableCertificateOwnerRef, "enable-certificate-owner-ref", defaultEnableCertificateOwnerRef, ""+
 		"Whether to set the certificate resource as an owner of secret where the tls certificate is stored. "+
 		"When this flag is enabled, the secret will be automatically removed when the certificate resource is deleted.")
+	fs.StringVar(&s.DefaultKeyAlgorithm, "default-key-algorithm", defaultKeyAlgorithm, ""+
+		"Default Key Algorithm for the certificate controller to inject when KeyAlgorithm and KeySize aren't specified for a certificate, "+
+		"this includes certificates created by the ingress-shim controller.")
+	fs.IntVar(&s.DefaultKeySize, "default-key-size", defaultKeySize, "" +
+		"Default Key Size for the certificate controller to inject when KeyAlgorithm and KeySize aren't specified for a certificate, "+
+		"this includes certificates created by the ingress-shim controller.")
+
 	fs.IntVar(&s.MaxConcurrentChallenges, "max-concurrent-challenges", defaultMaxConcurrentChallenges, ""+
 		"The maximum number of challenges that can be scheduled as 'processing' at once.")
 }
 
 func (o *ControllerOptions) Validate() error {
+	switch o.DefaultKeyAlgorithm {
+	case "", "rsa":
+		switch {
+		case o.DefaultKeySize == 0:
+		case 2048 <= o.DefaultKeySize && o.DefaultKeySize <= 8192:
+		default:
+			return fmt.Errorf("invalid default key size for rsa: %v", o.DefaultKeySize)
+		}
+	case "ecdsa":
+		switch o.DefaultKeySize {
+		case 0:
+		case 256:
+		case 384:
+		case 521:
+		default:
+			return fmt.Errorf("invalid default key size for ecdsa: %v", o.DefaultKeySize)
+		}
+	default:
+		return fmt.Errorf("invalid default key algorithm: %v", o.DefaultKeyAlgorithm)
+	}
```

Comment:
We already have logic similar to this here: https://github.com/jetstack/cert-manager/blob/master/pkg/util/pki/csr.go#L263-L266.

Is there any opportunity for code reuse here to avoid issues in future? i.e. some generalised validation function that validates a given keyAlgorithm & keySize combination is valid?

Reply:
I reused the method you pointed too with a slight modification. Let me know if you like it.